dbms/src/Flash/Mpp/MPPTunnelSet.cpp

```diff
@@ -63,83 +63,195 @@ void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
 }
 
-static inline void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t sz, bool is_local)
+template <typename Tunnel>
+static void BroadcastOrPassThroughWriteImpl(
+    std::vector<std::shared_ptr<Tunnel>> & tunnels,
+    size_t local_tunnel_cnt, // can be 0 for PassThrough writer
+    bool is_broadcast,
+    TrackedMppDataPacketPtr && tracked_packet,
+    TrackedMppDataPacketPtr && compressed_tracked_packet,
+    CompressionMethod compression_method)
 {
-    // statistic
-    GET_METRIC(tiflash_exchange_data_bytes, type_hash_original).Increment(original_size);
+    const size_t remote_tunnel_cnt = tunnels.size() - local_tunnel_cnt;
+    auto tracked_packet_bytes = tracked_packet ? tracked_packet->getPacket().ByteSizeLong() : 0;
+    auto compressed_tracked_packet_bytes = compressed_tracked_packet ? compressed_tracked_packet->getPacket().ByteSizeLong() : 0;
+    checkPacketSize(tracked_packet_bytes);
+    checkPacketSize(compressed_tracked_packet_bytes);
+
+    // TODO avoid copy packet for broadcast.
 
-    switch (method)
+    if (!compressed_tracked_packet)
     {
-    case CompressionMethod::NONE:
+        assert(tracked_packet);
+
+        compressed_tracked_packet_bytes = tracked_packet_bytes;
+
+        for (size_t i = 1; i < tunnels.size(); ++i)
+            tunnels[i]->write(tracked_packet->copy());
+        tunnels[0]->write(std::move(tracked_packet));
+    }
+    else
     {
-        if (is_local)
-        {
-            GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_local).Increment(sz);
-        }
-        else
+        for (size_t i = 0, local_cnt = 0, remote_cnt = 0; i < tunnels.size(); ++i)
         {
-            GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_remote).Increment(sz);
+            if (IsLocalTunnel(tunnels[i]))
+            {
+                local_cnt++;
+                if (local_cnt == local_tunnel_cnt)
+                    tunnels[i]->write(std::move(tracked_packet));
+                else
+                    tunnels[i]->write(tracked_packet->copy()); // NOLINT
+            }
+            else
+            {
+                remote_cnt++;
+                if (remote_cnt == remote_tunnel_cnt)
+                    tunnels[i]->write(std::move(compressed_tracked_packet));
+                else
+                    tunnels[i]->write(compressed_tracked_packet->copy()); // NOLINT
+            }
         }
-        break;
     }
-    case CompressionMethod::LZ4:
+
+    if (is_broadcast)
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_lz4_compression).Increment(sz);
-        break;
+        UPDATE_EXCHANGE_MATRIC(broadcast, CompressionMethod::NONE, local_tunnel_cnt * tracked_packet_bytes, local_tunnel_cnt * tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(broadcast, compression_method, remote_tunnel_cnt * tracked_packet_bytes, remote_tunnel_cnt * compressed_tracked_packet_bytes, false);
```

Comment:
I think `tracked_packet_bytes` can be 0 here if there is no local tunnel

Reply:
Fixed. I have use another parameter `size_t ori_tracked_packet_bytes`.